### ISSUE TITLE
Assertion failed in test/test-pipe-set-non-blocking.c on line 51: n == 0

### ISSUE THREAD

**Issue by xi3** [libuv/libuv#248]
```

make  check-TESTS
make[1]: Entering directory '/home/xi/download/libuv-1.4.2'
[%   0|+   0|-   0|T   0|S   0]: platform_output
Output from process `platform_output`:
uv_get_process_title: /home/xi/download/libuv-1.4.2/test/.libs/lt-run-tests
uv_cwd: /home/xi/download/libuv-1.4.2
uv_resident_set_memory: 749568
uv_uptime: 491608.000000
uv_getrusage:
  user: 0 sec 0 microsec
  system: 0 sec 1026 microsec
uv_cpu_info:
  model: QEMU Virtual CPU
  speed: 2099
  times.sys: 15097100
  times.user: 22070600
  times.idle: 4858393700
  times.irq: 0
  times.nice: 1556400
uv_interface_addresses:
  name: lo
  internal: 1
  physical address: 00:00:00:00:00:00
  address: 127.0.0.1
  netmask: 255.0.0.0
  name: eth0
  internal: 0
  physical address: 52:54:42:80:c0:22
  address: 10.70.56.196
  netmask: 255.255.255.0
  name: lo
  internal: 1
  physical address: 00:00:00:00:00:00
  address: ::1
  netmask: ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff
  name: eth0
  internal: 0
  physical address: 52:54:42:80:c0:22
  address: fe80::5054:42ff:fe80:c022
  netmask: ffff:ffff:ffff:ffff::
=============================================================
[%  10|+  26|-   0|T   0|S   0]: pipe_set_non_blocking
`pipe_set_non_blocking` failed: exit code 134
Output from process `pipe_set_non_blocking`:
Assertion failed in test/test-pipe-set-non-blocking.c on line 51: n == 0
=============================================================
[% 100|+ 249|-   1|T   0|S   0]: Done.
FAIL: test/run-tests
======================================================
1 of 1 test failed
Please report to https://github.com/libuv/libuv/issues
======================================================
Makefile:3835: recipe for target 'check-TESTS' failed
make[1]: *** [check-TESTS] Error 1
make[1]: Leaving directory '/home/xi/download/libuv-1.4.2'
Makefile:4091: recipe for target 'check-am' failed
make: *** [check-am] Error 2

···

**Comment by saghul** [comment]
What platform are you running? Does the test consistently fail on different runs?

**Comment by txdv** [comment]
QEMU Virtual CPU

What OS are you running though?

**Comment by sewenew** [comment]
I got the same problem with v1.4.2. My platform is 32bit Linux (Fedora 13), and the test consistently fail. Any suggestion? Thank you!

**Comment by saghul** [comment]
@bnoordhuis any idea?

**Comment by bnoordhuis** [comment]
I speculate that the read() system call is failing with EPIPE.  Does this patch help?

```diff
diff --git a/test/test-pipe-set-non-blocking.c b/test/test-pipe-set-non-blocking.c
index 5cf2c19..f6993d3 100644
--- a/test/test-pipe-set-non-blocking.c
+++ b/test/test-pipe-set-non-blocking.c
@@ -48,7 +48,7 @@ static void thread_main(void* arg) {
     n = read(ctx->fd, buf, sizeof(buf));
   while (n > 0 || (n == -1 && errno == EINTR));
 
-  ASSERT(n == 0);
+  ASSERT(n == 0 || (n == -1 && errno == EPIPE));
 }
 
 TEST_IMPL(pipe_set_non_blocking) {
```

If not, can you strace the test and post the log file here?

```
# you may have to set LD_LIBRARY_PATH
$ strace -o trace.log -fe read test/.libs/lt-run-tests pipe_set_non_blocking pipe_set_non_blocking
```

**Comment by sewenew** [comment]
The patch doesn't work. I checked the value of 'n' and 'errno': n == 1 && errno == EBADF.

It seems the main thread runs much faster, and closes fd[1] (line 91) when the sub-thread is still reading from fd[1].

**Comment by sewenew** [comment]
trace.log: 
1573  read(3, "\177ELF\1\1\1\3\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0@K\0\0004\0\0\0"..., 512) = 512
1573  read(3, "\177ELF\1\1\1\3\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0\200\370\223\0004\0\0\0"..., 512) = 512
1573  read(3, "\177ELF\1\1\1\3\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0\200\234\220\0004\0\0\0"..., 512) = 512
1573  read(3, "\177ELF\1\1\1\3\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0p\201l\0074\0\0\0"..., 512) = 512
1573  read(3, "\177ELF\1\1\1\3\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0`*\222\0004\0\0\0"..., 512) = 512
1573  read(3, "\177ELF\1\1\1\3\0\0\0\0\0\0\0\0\3\0\3\0\1\0\0\0 \336x\0004\0\0\0"..., 512) = 512
1574  read(11, "................................"..., 4096) = 4096
// repeat 2560 times...
1574  read(11, "................................"..., 4096) = 4096
1574  read(11, "", 4096)                = 0

**Comment by bnoordhuis** [comment]
Oh right, I think I've spotted the error.  Can you confirm that this patch fixes it?

```diff
diff --git a/test/test-pipe-set-non-blocking.c b/test/test-pipe-set-non-blocking.c
index 5cf2c19..fcc9fc0 100644
--- a/test/test-pipe-set-non-blocking.c
+++ b/test/test-pipe-set-non-blocking.c
@@ -88,8 +88,8 @@ TEST_IMPL(pipe_set_non_blocking) {
   uv_close((uv_handle_t*) &pipe_handle, NULL);
   ASSERT(0 == uv_run(uv_default_loop(), UV_RUN_DEFAULT));
 
-  ASSERT(0 == close(fd[1]));  /* fd[0] is closed by uv_close(). */
   ASSERT(0 == uv_thread_join(&thread));
+  ASSERT(0 == close(fd[1]));  /* fd[0] is closed by uv_close(). */
   uv_barrier_destroy(&ctx.barrier);
 
   MAKE_VALGRIND_HAPPY();
```

**Comment by sewenew** [comment]
It works! Thank you!

**Comment by bnoordhuis** [comment]
Fixed in 531b085.

**Comment by bnoordhuis** [comment]
And thanks for testing. :-)
